halo2_proofs/src/plonk/evaluation.rs

```diff
@@ -517,6 +555,68 @@ impl<C: CurveAffine> Evaluator<C> {
                     }
                 });
             }
+
+            // Shuffle constraints
+            for (n, shuffle) in shuffles.iter().enumerate() {
+                let product_coset = pk.vk.domain.coeff_to_extended(shuffle.product_poly.clone());
+
+                // Shuffle constraints
+                parallelize(&mut values, |values, start| {
+                    let input_evaluator = &self.shuffles[2 * n];
+                    let shuffle_evaluator = &self.shuffles[2 * n + 1];
+                    let mut eval_data_input = shuffle_evaluator.instance();
+                    let mut eval_data_shuffle = shuffle_evaluator.instance();
+                    for (i, value) in values.iter_mut().enumerate() {
+                        let idx = start + i;
+
+                        let input_value = input_evaluator.evaluate(
+                            &mut eval_data_input,
+                            fixed,
+                            advice,
+                            instance,
+                            challenges,
+                            &beta,
+                            &gamma,
+                            &theta,
+                            &y,
+                            &C::ScalarExt::ZERO,
+                            idx,
+                            rot_scale,
+                            isize,
+                        );
+
+                        let shuffle_value = shuffle_evaluator.evaluate(
+                            &mut eval_data_shuffle,
+                            fixed,
+                            advice,
+                            instance,
+                            challenges,
+                            &beta,
+                            &gamma,
+                            &theta,
+                            &y,
+                            &C::ScalarExt::ZERO,
+                            idx,
+                            rot_scale,
+                            isize,
+                        );
+
+                        let r_next = get_rotation_idx(idx, 1, rot_scale, isize);
+
+                        // l_0(X) * (1 - z(X)) = 0
+                        *value = *value * y + ((one - product_coset[idx]) * l0[idx]);
+                        // l_last(X) * (z(X)^2 - z(X)) = 0
+                        *value = *value * y
+                            + ((product_coset[idx] * product_coset[idx] - product_coset[idx])
+                                * l_last[idx]);
+                        // (1 - (l_last(X) + l_blind(X))) * (z(\omega X) (s(X) + \gamma) - z(X) (a(X) + \gamma) = 0
```

Comment:
```suggestion
                        // (1 - (l_last(X) + l_blind(X))) * (z(\omega X) (s(X) + \gamma) - z(X) (a(X) + \gamma)) = 0
```
There's a missing closing parenthesis.